matchtree.go

```diff
@@ -1010,6 +1011,77 @@ func (d *indexData) newMatchTree(q query.Q) (matchTree, error) {
 	return nil, nil
 }
 
+func (d *indexData) bloomReject(q query.Q) bool {
```

Comment:
you are not handling other types which can have children, look at the implementation of `query.Map`. I recommend doing bloom rejection inside of `indexData.simplify`

Reply:
query.Not(query that's true because of bloom filter) doesn't become false, because it could be a false positive.

When does the frontend send type:  queries? Any unhandled here are safe, since returning 'false' just means 'don't skip this shard'.

Reply:
> query.Not(query that's true because of bloom filter) doesn't become false, because it could be a false positive.

Yup. But you should only replace a Q with `FALSE` IFF you get the bloom filter rejecting it. IE do not put in `TRUE`. Then after the simplify run if the output is `FALSE` you know you can skip. That will handle NOT,OR,AND correctly.

> When does the frontend send type: queries? Any unhandled here are safe, since returning 'false' just means 'don't skip this shard'.

It is currently used for `repoHasFile:` type queries, so rarely. https://sourcegraph.com/search?q=context:global+%28zoekt%7Cquery%29%5C.Type+repo:%5Egithub%5C.com/sourcegraph/sourcegraph%24+lang:go+&patternType=regexp

Ok, great, the fallback sounds good.

I need to review this implementation more closely to understand if it is correct. However, if you just do this as part of `indexdata.simplify` there is no need to review closely. Given we already do the simplify step, it feels like it is better to do it there? Additionally doing it in simplify will help prune out Q's that will nevermatch. eg `FOO or BAR` where `BAR` fails the bloom filter will simplify into `FOO` before we create the matchtree.

I see this change does this check before `simplify`. Do you have any measurements showing that avoiding allocations in that will help? Should we consider a version of simplify which is like `query.Map`, but just takes into account constant folding? That way we can do it without allocating to the heap like in your bloomReject func.